include/zephyr/drivers/intc.h

```diff
@@ -0,0 +1,462 @@
+		.priority = INTC_DT_GET_IRQ_PRIORITY_BY_NAME(node_id, name),	\
+	}
+
+/**
+ * @brief Get IRQ devicetree specifier by node identifier and name if IRQ exists
+ * @note Identical to INTC_DT_GET_IRQ_SPEC_BY_NAME() if IRQ exists
+ * @see INTC_DT_GET_IRQ_SPEC_BY_NAME()
+ */
+#define INTC_DT_GET_OPT_IRQ_SPEC_BY_NAME(node_id, name)				\
+	COND_CODE_1(								\
+		DT_IRQ_HAS_NAME(node_id, name),					\
+		(INTC_DT_GET_IRQ_SPEC_BY_NAME(node_id, name)),			\
+		({})								\
+	)
+
+/**
+ * @brief Get the first IRQ devicetree specifier by node identifier
+ * @note Identical to INTC_DT_SPEC_GET_BY_IDX(node_id, 0)
+ * @see INTC_DT_SPEC_GET_BY_IDX()
+ */
+#define INTC_DT_GET_IRQ_SPEC(node_id) \
+	INTC_DT_GET_IRQ_SPEC_BY_IDX(node_id, 0)
+
+/**
+ * @brief Get first IRQ devicetree specifier by node identifier if IRQ exists
+ * @note Identical to INTC_DT_SPEC_GET_OPT_BY_IDX(node_id, 0)
+ * @see INTC_DT_SPEC_GET_OPT_BY_IDX()
+ */
+#define INTC_DT_GET_OPT_IRQ_SPEC(node_id) \
+	INTC_DT_GET_OPT_IRQ_SPEC_BY_IDX(node_id, 0)
+
+/**
+ * @brief Get IRQ devicetree specifier by DT_DRV_COMPAT instance and idx
+ * @see INTC_DT_SPEC_GET_BY_IDX()
+ */
+#define INTC_DT_INST_GET_IRQ_SPEC_BY_IDX(inst, idx) \
+	INTC_DT_GET_IRQ_SPEC_BY_IDX(DT_DRV_INST(inst), idx)
+
+/**
+ * @brief Get IRQ devicetree specifier by DT_DRV_COMPAT instance and idx if IRQ exists
+ * @see INTC_DT_SPEC_GET_OPT_BY_IDX()
+ */
+#define INTC_DT_INST_GET_OPT_SPEC_BY_IDX(inst, idx) \
+	INTC_DT_GET_OPT_IRQ_SPEC_BY_IDX(DT_DRV_INST(inst), idx)
+
+/**
+ * @brief Get IRQ devicetree specifier by DT_DRV_COMPAT instance and name
+ * @see INTC_DT_SPEC_GET_BY_NAME()
+ */
+#define INTC_DT_INST_GET_IRQ_SPEC_BY_NAME(inst, name) \
+	INTC_DT_GET_IRQ_SPEC_BY_NAME(DT_DRV_INST(inst), name)
+
+/**
+ * @brief Get IRQ devicetree specifier by DT_DRV_COMPAT instance and name if IRQ exists
+ * @see INTC_DT_SPEC_GET_OPT_BY_NAME()
+ */
+#define INTC_DT_INST_GET_OPT_IRQ_SPEC_BY_NAME(inst, name) \
+	INTC_DT_GET_OPT_IRQ_SPEC_BY_NAME(DT_DRV_INST(inst), name)
+
+/**
+ * @brief Get the first IRQ devicetree specifier by DT_DRV_COMPAT instance
+ * @see INTC_DT_SPEC_GET()
+ */
+#define INTC_DT_INST_GET_IRQ_SPEC(inst) \
+	INTC_DT_GET_IRQ_SPEC(DT_DRV_INST(inst))
+
+/**
+ * @brief Get first IRQ devicetree specifier by DT_DRV_COMPAT instance if IRQ exists
+ * @see INTC_DT_SPEC_GET_OPT()
+ */
+#define INTC_DT_INST_GET_OPT_IRQ_SPEC(inst) \
+	INTC_DT_GET_OPT_IRQ_SPEC(DT_DRV_INST(inst))
+
+/** Interrupt request handle forward declaration */
+struct intc_irq;
+
+/** Interrupt request handler */
+typedef void (*intc_irq_handler_t)(struct intc_irq *irq);
+
+/** @cond INTERNAL_HIDDEN */
+
+/** Interrupt request specification */
+struct intc_irq_spec {
+	/** Interrupt line's interrupt controller */
+	const struct device *intc;
+	/** Interrupt controller's interrupt line number  */
+	uint16_t irq;
+	/** Interrupt request priority */
+	uint16_t priority;
+	/** Interrupt request flags */
+	uint32_t flags;
```

Comment:
the priority and irq number are integers, what will probably never exceed an 65365, but do exceed 255, where the flags is a bitfield, so it should be as large as we can provide :)

That said, I may update flags to be a size_t, but it does seem MSI fields can fit inside a uint32 https://i.stack.imgur.com/RwenJ.jpg

Reply:
Seems like this is fine for MSI.

Anyway, the implementation of MSI/MSI-X is currently very much arch specific, that is: x86 specific. So there won't be any collision with such generic intc API. 

It could change someday, since interrupt management implementation in x86 is messy to say the least (you always need to allocate an irq to get a vector for instance, that does not make much sense in case of MSI-X or multi-vector MSI).